### ISSUE TITLE
Long delay between iterations when evaluating model repeatedly in a for loop

### ISSUE THREAD

**Issue by 2gunsu** [ultralytics/ultralytics#20763]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I am experiencing a significant delay after each evaluation iteration completes and before the next one begins when running multiple evaluation passes on the same model using a for loop.

The example code is as follows.
```python
model = YOLO(model=weight_file, verbose=False)

for iou_thres in iou_thres_list:
    results = model.val(
        data=data_yaml, 
        imgsz=image_size, 
        conf=val_conf, 
        iou=iou_thres, 
        augment=False, 
        verbose=False)

        .....

```

What could be the potential causes for such a delay?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @2gunsu, thank you for reaching out and providing your code example! 🚀 This is an automated response to help you get the fastest possible support. An Ultralytics engineer will review your issue and assist you soon.

We recommend visiting the [Docs](https://docs.ultralytics.com/) for comprehensive guidance, with many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples that address common questions.

If this is a 🐛 Bug Report, please include a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE) to help us investigate the issue efficiently.

For custom training or evaluation ❓ questions, providing detailed information—such as dataset samples, logs, and a description of your environment—will help us diagnose the cause more quickly. Also, ensure you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community:
- Real-time chat: [Discord](https://discord.com/invite/ultralytics) 🎧
- In-depth discussions: [Discourse](https://community.ultralytics.com/)
- Community threads: [Subreddit](https://reddit.com/r/Ultralytics)

## Upgrade

Please ensure you are using the latest `ultralytics` package and dependencies in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/). Upgrade using:

```bash
pip install -U ultralytics
```

See all package requirements [here](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml).

## Environments

YOLO runs in a variety of verified environments (all dependencies, including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu.

Thank you for your patience!

**Comment by glenn-jocher** [comment]
The delay you're experiencing is likely due to the validator cleanup and reinitialization process that occurs between each `model.val()` call. When running multiple evaluations in a loop, the validator creates new dataloaders, reinitializes metrics, and performs other setup operations for each iteration.

To reduce this overhead, you could consider creating a single validator instance and reusing it, or if you're specifically testing different IoU thresholds, you might want to run validation once and then post-process the results at different IoU levels. The [validation mode documentation](https://docs.ultralytics.com/modes/val/) provides more details on the validation process and available parameters.

**Comment by 2gunsu** [comment]
Thank you for your kind answer! :)
